File: be/api/hospital.py
"""병원 상세 API 라우터.

스펙: .claude/docs/API-FE-BE.md > 엔드포인트 > 2. 병원 상세
응답: 상세 페이지 9개 영역에 직접 매핑.
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from fastapi.responses import Response

from be.adapters.dynamo_adapter import DynamoAdapter
from be.adapters.s3_adapter import S3Adapter
from be.core.feedback import compute_feedback_stats
from shared.etc_category import display_specialty

router = APIRouter(prefix="/api/hospitals", tags=["hospitals"])
db = DynamoAdapter()
s3 = S3Adapter()


@router.get("/{hospital_id}/thumbnail")
def get_hospital_thumbnail(hospital_id: str):
    """병원 대표 이미지(홈페이지 스크린샷) 스트리밍.

    S3 크롤 버킷이 public-access 차단이라 외부 직접노출이 안 돼 BE 가 중계한다.
    카카오/네이버 대표사진은 외부 CDN URL 을 thumbnail_url 에 직접 넣으므로 이 경로를
    안 탄다 — 이 엔드포인트는 우리가 찍어 S3 에 올린 스크린샷 전용. 없으면 404
    (FE HospitalThumbnail 이 onError 로 회색 플레이스홀더 폴백).
    """
    data = s3.load_thumbnail(hospital_id)
    if not data:
        raise HTTPException(status_code=404, detail="thumbnail not found")
    return Response(
        content=data,
        media_type="image/jpeg",
        headers={"Cache-Control": "public, max-age=86400"},
    )


@router.get("/{hospital_id}")
def get_hospital_detail(hospital_id: str):
    """
    병원 상세 페이지 — 9개 영역 데이터 반환.

    ① AI 통합 설명 (ai_description)
    ② 핵심 진료 정보 (services, excluded_services, equipment, prices)
    ③ 의료진 정보 (doctors)
    ④ 신뢰도·근거 (confidence, detailed_signals)
    ⑤ 기본 운영 정보 (location, operating_hours, contact)
    ⑥ 사용자 피드백 (feedback_stats)
    ⑦ 분류 변경 이력 (recent_changes)
    ⑧ 관련 병원 추천 (related_hospitals)
    ⑨ 메타 정보 (metadata)
    """
    meta = db.load_hospital_meta(hospital_id)
    if not meta:
        raise HTTPException(
            status_code=404,
            detail={"error": {"code": "NOT_FOUND", "message": "병원을 찾을 수 없습니다"}},
        )

    classification = db.load_classification(hospital_id)
    description = db.load_description(hospital_id)
    services_and_doctors = db.load_services_and_doctors(hospital_id)
    related = db.load_related_hospitals(hospital_id)
    recent_changes = db.load_recent_changes(hospital_id, limit=2)

    # 심평원 공공 데이터 (③ 의료진 보강·비급여 신규 영역)
    public_doctors = db.load_public_doctors(hospital_id)
    public_nonpay = db.load_public_nonpay(hospital_id)

    # ⑥ 피드백 통계
    feedback_list = db.get_feedback_for_hospital(hospital_id)
    feedback_stats = compute_feedback_stats(feedback_list)

    # 응답 조립 — API-FE-BE.md 스펙 기준
    return {
        "data": {
            "hospital_id": meta.hospital_id,
            "name": meta.name,
            "standard_specialty": classification.standard_specialty if classification else "",
            "etc_subcategory": (
                display_specialty(classification.standard_specialty, classification.primary_focus)
                if classification else ""
            ),
            "primary_focus": classification.primary_focus if classification else [],
            "confidence": classification.confidence.model_dump() if classification else None,
            # 헤드라이너 히어로 + 카드 좌측 썸네일 동일 출처. None 이면 그라데이션 플레이스홀더.
            "thumbnail_url": meta.thumbnail_url,
            "location": meta.location.model_dump() if meta.location else None,
            "website_url": meta.contact.website_url,
            "one_line_summary": description.one_line_summary if description else "",

            # ① AI 통합 설명
            "ai_description": description.model_dump() if description else None,

            # ② 핵심 진료 정보 (스펙 형태로 어댑팅)
            "services": [s.model_dump() for s in services_and_doctors.services] if services_and_doctors else [],
            "excluded_services": [s.model_dump() for s in services_and_doctors.excluded_services] if services_and_doctors else [],
            "equipment": [_adapt_equipment(e) for e in services_and_doctors.equipment] if services_and_doctors else [],
            "prices": [_adapt_price(p) for p in services_and_doctors.prices] if services_and_doctors else [],

            # ③ 의료진 정보 + 심평원 신고 기준 전문의 수
            "doctors": [_adapt_doctor(d) for d in services_and_doctors.doctors] if services_and_doctors else [],
            # 심평원 신고 기준 과목별 전문의 수 {"피부과": 2, ...}. 출처: 심평원 공식 신고(public_data).
            # "이 병원이 잘 본다" 아닌 "심평원 신고 기준 N명" 형식으로 FE 렌더 필요(의료법 주체 명시).
            "specialists_by_dept": public_doctors.get("specialists_by_dept", {}),
            # 총 의사 수(getDtlInfo2.7). None 이면 미확인.
            "total_doctors": public_doctors.get("total_doctors"),

            # (신규) 심평원 신고 비급여 항목 — 의료법 제45조의2 공식 신고 사실 그대로 노출.
            # "병원이 신고한 비급여 항목"(주체 명시). 분류 신호 편입 금지, 표시·필터 전용.
            "nonpay_items": [item.model_dump() for item in public_nonpay],

            # ④ 신뢰도·근거
            "detailed_signals": _adapt_detailed_signals(classification, meta.contact.website_url if meta.contact else None),

            # ⑤ 기본 운영 정보 — operating_hours 는 구조화 미보유라 null(FE 가 "정보 없음")
            "operating_hours": None,
            "contact": _adapt_contact(meta.contact),

            # ⑥ 사용자 피드백
            "feedback_stats": feedback_stats.model_dump(),

            # ⑦ 분류 변경 이력
            "recent_changes": [c.model_dump() for c in recent_changes],

            # ⑧ 관련 병원 추천
            "related_hospitals": [_adapt_related(r) for r in related],

            # ⑨ 메타 정보
            "metadata": {
                "last_updated_at": classification.classified_at.isoformat() if classification else None,
                "data_sources": ["public_registry"],
                "data_completeness": _calc_completeness(classification, description, services_and_doctors),
                "warning": "정보 부족 — 직접 병원에 문의 권장" if not classification else None,
            },
        }
    }


# ---------------------------------------------------------------------------
# 응답 어댑터 — raw Pydantic 모델 → API-FE-BE.md 스펙 형태 (FE 가 기대하는 shape).
# BE 모델이 안 가진 필드(source_text·sample_image_urls·career·주차 등)는 null/[]/false
# 로 채운다(미수집 = graceful 공란). 스펙이 합의된 계약이므로 BE 가 거기 맞춘다.
# ---------------------------------------------------------------------------

def _adapt_detailed_signals(classification, website_url: str | None) -> dict | None:
    if not classification:
        return None
    ds = classification.detailed_signals
    # self_claim 은 아래에서 classification.primary_focus 로 대체 노출하므로 ds.self_claim 미사용.
    v, blog, rev = ds.vision, ds.blog, ds.reviews
    return {
        "self_claim": {
            # 정제된 주력(primary_focus)을 자칭 컨셉으로 노출. raw sc.keywords 는 자기 사이트의
            # 블로그/FAQ 문맥어(예: 탈모약 부작용 FAQ의 '당뇨·기형아·분만')까지 섞여 오인 소지 →
            # 빈도·교차검증 거친 primary_focus 로 대체(분류 노이즈가 상세 화면에 그대로 노출되는 것 방지).
            "extracted_keywords": list(classification.primary_focus or []),
            "source_text": "",  # 자칭 원문은 임베딩 전용·미저장
            "source_url": website_url or "",
        },
        "vision": (
            {
                "detected_devices": list(v.detected_devices or []),
                "image_distribution": dict(v.image_categories or {}),
                "sample_image_urls": [],  # 샘플 이미지 URL 미수집
            }
            if v
            else None
        ),
        "blog": {
            "top_topics": [
                {"topic": t, "frequency": int(f)}
                for t, f in list((blog.keyword_frequency or {}).items())[:10]
            ],
            "total_posts": blog.total_posts,
        },
        "reviews": {
            "review_count": rev.total_reviews,
            "top_keywords": list((rev.keyword_frequency or {}).keys())[:15],
        },
    }


_EQUIP_SOURCE_MAP = {"public_data": "public_registry", "vision": "vision"}


def _adapt_equipment(e) -> dict:
    return {
        "name": e.name,
        "available": True,
        "source": _EQUIP_SOURCE_MAP.get(e.source, "self_claim"),
        "source_url": None,
    }


def _adapt_price(p) -> dict:
    return {"service_name": p.service_name, "price_range": p.price_text, "source_url": "", "last_seen": ""}


def _adapt_doctor(d) -> dict:
    return {
        "name": d.name,
        "position": d.specialty or "",
        "specialty_certifications": list(d.qualifications or []),
        "sub_specialty": d.sub_specialty,
        "career": [],
        "primary_focus": None,
        "source_url": None,
    }


def _adapt_contact(c) -> dict | None:
    if not c:
        return None
    methods: list[str] = []
    if c.reservation_url:
        methods.append("online")
    if c.phone:
        methods.append("phone")
    return {
        "phone": c.phone or "",
        "homepage_url": c.website_url,
        "parking_available": False,
        "appointment_methods": methods,
    }


def _adapt_related(r) -> dict:
    d = r.model_dump()
    # 관련 병원 카드 — 이름·대표 이미지를 META 에서 조인(≤5건). find_related_hospitals 는
    # name="" 로 두고 "BE 가 조인"하기로 한 계약이라, 여기서 이름을 안 채우면 FE 가 빈 이름
    # 대신 hospital_id(raw) 를 노출한다(관측된 'ID 가 왜 여기' 버그).
    rel_meta = db.load_hospital_meta(r.hospital_id)
    if rel_meta:
        if not (d.get("name") or "").strip():
            d["name"] = rel_meta.name
        d["thumbnail_url"] = rel_meta.thumbnail_url
    else:
        d["thumbnail_url"] = None
    # primary_focus literal 따옴표 정규화(적재 데이터 오염 방지).
    if isinstance(d.get("primary_focus"), list):
        d["primary_focus"] = [
            str(f).strip().strip('"').strip("'").strip()
            for f in d["primary_focus"] if str(f).strip()
        ]
    return d


def _calc_completeness(classification, description, services_and_doctors) -> float:
    """9개 영역 중 채워진 비율 계산."""
    filled = 1  # meta는 항상 있음 (여기까지 왔으니까)
    if classification:
        filled += 2  # ② 일부 + ④
    if description:
        filled += 1  # ①
    if services_and_doctors:
        filled += 2  # ② 나머지 + ③
    # ⑤ 기본 운영 정보는 meta에 포함
    filled += 1
    # ⑥⑦⑧은 데이터 있으면 카운트
    return round(filled / 9, 2)

File: fe/src/mocks/hospital.ts
import type { HospitalDetail } from "@/types/domain";

// API-FE-BE.md "병원 상세" 응답을 기반으로 한 데모 Mock.
//
// 데이터는 9영역 위계 다듬기 라운드에 맞춰 풍성하게 확장했다:
//   - 의사 3명: 원장(아토피·습진 세부 전공) / 부원장(여드름 케어 + 의사별 primary_focus) / 진료의(일반)
//     · API 명세상 primary_focus 는 "의사별로 다른 경우만" 채움 → 한 명만 채워서 두 패턴 시각화
//   - 서비스: 진료과목별 4 카테고리 모두 등장
//   - excluded_services / equipment: 헛걸음 방지 시연 케이스
//   - 비급여 가격 다중 항목
//   - related_hospitals: same_focus 2 + fills_gap 2
//   - recent_changes 2건 (재분류 + 초기 분류)
//
// BE 연동 전까진 이 Mock 으로 9영역 화면 골격을 채운다.
export const mockHospital: HospitalDetail = {
  hospital_id: "h_abc123",
  name: "○○피부과의원",
  standard_specialty: "피부과",
  primary_focus: ["일반 진료 (아토피·여드름)"],
  confidence: {
    score: 92,
    level: "확실",
    signals: { self_claim: 35, vision: 28, blog: 17, reviews: 12 },
  },
  location: {
    address: "서울특별시 마포구 공덕동 1-2",
    sido: "서울특별시",
    sigungu: "마포구",
    dong: "공덕동",
    lat: 37.5443,
    lng: 126.951,
  },
  website_url: "https://example.com",
  one_line_summary: "일반 피부 진료 중심, 미용 시술은 거의 안 하는 동네 의원",

  ai_description: {
    headline: "○○피부과는 일반 피부 진료 중심의 동네 의원입니다.",
    paragraphs: [
      {
        text: "홈페이지 메인 화면에서 아토피·여드름·습진 같은 일반 피부질환을 가장 먼저 안내하고 있으며, 시술 사진 80%가 일반 진료 케이스(피부 발진·습진·여드름)고 미용 시술 사진은 18%로 적습니다. 블로그 글 50건 중 아토피 관련 글이 34%, 여드름 관련 글이 21%로 가장 많고, 미용 시술 관련 글은 5건뿐입니다.",
        citations: ["self_claim", "vision", "blog"],
      },
      {
        text: "실제 방문 후기에서도 '친절한 아토피 상담', '꼼꼼한 여드름 치료' 같은 키워드가 자주 등장합니다. 다만 사마귀 냉동치료기·점 제거 레이저 같은 시술 장비는 보유하고 있지 않은 것으로 보이므로, 미용 목적이라면 다른 병원을 권합니다.",
        citations: ["reviews", "vision"],
      },
    ],
    generated_at: "2026-04-12T08:00:00Z",
    generator_model: "anthropic.claude-sonnet-4-5-20250929-v1:0",
  },

  services: [
    {
      name: "아토피",
      category: "general",
      source_signals: ["self_claim", "blog", "reviews"],
    },
    {
      name: "여드름",
      category: "general",
      source_signals: ["self_claim", "blog", "reviews"],
    },
    {
      name: "습진",
      category: "general",
      source_signals: ["self_claim", "blog"],
    },
    {
      name: "지루성 피부염",
      category: "general",
      source_signals: ["self_claim", "blog"],
    },
    {
      name: "점 빼기",
      category: "cosmetic",
      source_signals: ["self_claim"],
    },
    {
      name: "피부 검사",
      category: "exam",
      source_signals: ["self_claim", "vision"],
    },
  ],

  excluded_services: [
    {
      name: "사마귀 냉동치료",
      reason: "no_equipment",
      alternative_hospital_ids: ["h_def456", "h_ghi789"],
    },
    {
      name: "M자 탈모 처방",
      reason: "no_mention",
      alternative_hospital_ids: ["h_jkl012"],
    },
    {
      name: "레이저 토닝",
      reason: "low_signal",
      alternative_hospital_ids: [],
    },
  ],

  equipment: [
    {
      name: "더모스코프",
      available: true,
      source: "vision",
      source_url: "https://example.com/about",
    },
    {
      name: "패치 테스트 키트",
      available: true,
      source: "self_claim",
      source_url: "https://example.com/about",
    },
    {
      name: "사마귀 냉동치료기",
      available: false,
      source: "vision",
      source_url: null,
    },
    {
      name: "점 제거 레이저",
      available: false,
      source: "vision",
      source_url: null,
    },
  ],

  prices: [
    {
      service_name: "점 빼기 (1mm 기준)",
      price_range: "5만원~",
      source_url: "https://example.com/price",
      last_seen: "2026-04-12T08:00:00Z",
    },
    {
      service_name: "패치 테스트",
      price_range: "8만원",
      source_url: "https://example.com/price",
      last_seen: "2026-04-12T08:00:00Z",
    },
    {
      service_name: "더모스코프 진단",
      price_range: "3만원",
      source_url: "https://example.com/price",
      last_seen: "2026-04-12T08:00:00Z",
    },
  ],

  doctors: [
    {
      name: "김가영",
      position: "원장 · 대표원장",
      specialty_certifications: ["피부과 전문의", "대한피부과학회 정회원"],
      sub_specialty: "아토피·습진",
      career: [
        "서울대학교 의과대학 졸업",
        "서울대학교병원 피부과 인턴·레지던트 수료",
        "대한피부과학회 학술대회 발표 (2022)",
      ],
      primary_focus: null,
      source_url: "https://example.com/doctor/1",
    },
    {
      name: "이서준",
      position: "부원장",
      specialty_certifications: ["피부과 전문의"],
      sub_specialty: "여드름·지루성 피부염",
      career: [
        "연세대학교 의과대학 졸업",
        "세브란스병원 피부과 수련",
        "○○피부과 합류 (2023)",
      ],
      // 이 의사만 진료 분야가 다른 케이스 — 의사별 primary_focus 가 전체와
      // 다르게 설정될 수 있다는 API 명세를 시각화
      primary_focus: ["여드름 케어", "지루성 피부염"],
      source_url: "https://example.com/doctor/2",
    },
    {
      name: "박지유",
      position: "진료의",
      specialty_certifications: ["피부과 전문의"],
      sub_specialty: null,
      career: ["고려대학교 의과대학 졸업", "고대안암병원 피부과 수련"],
      primary_focus: null,
      source_url: null,
    },
  ],

  detailed_signals: {
    self_claim: {
      extracted_keywords: ["아토피", "여드름", "습진", "지루성 피부염"],
      source_text:
        "본원은 일반 피부 진료를 중심으로, 아토피·여드름·습진 등 만성 피부 질환의 꾸준한 관리를 중요하게 생각합니다. 미용 시술보다는 환자분의 피부 상태를 오래 지켜보며 치료 방향을 잡는 진료를 지향합니다.",
      source_url: "https://example.com/about",
    },
    vision: {
      detected_devices: ["더모스코프", "패치 테스트 키트"],
      image_distribution: {
        "일반 진료": 0.78,
        "미용 시술": 0.18,
        기타: 0.04,
      },
      sample_image_urls: [
        "https://example.com/img1.jpg",
        "https://example.com/img2.jpg",
      ],
    },
    blog: {
      top_topics: [
        { topic: "아토피", frequency: 0.34 },
        { topic: "여드름", frequency: 0.21 },
        { topic: "습진", frequency: 0.12 },
        { topic: "지루성 피부염", frequency: 0.08 },
      ],
      total_posts: 50,
    },
    reviews: {
      review_count: 142,
      top_keywords: ["친절", "아토피", "여드름", "꼼꼼", "설명 자세함"],
    },
  },

  operating_hours: {
    weekday: {
      open: "09:00",
      close: "18:00",
      lunch_start: "13:00",
      lunch_end: "14:00",
    },
    saturday: {
      open: "09:00",
      close: "13:00",
      lunch_start: null,
      lunch_end: null,
    },
    sunday: null,
    night_clinic: false,
    holiday_clinic: false,
  },

  contact: {
    phone: "02-1234-5678",
    homepage_url: "https://example.com",
    parking_available: true,
    appointment_methods: ["walk_in", "phone"],
  },

  feedback_stats: {
    total_count: 145,
    agree_count: 126,
    disagree_count: 19,
    agree_ratio: 0.87,
    last_feedback_at: "2026-05-15T14:22:00Z",
  },

  recent_changes: [
    {
      changed_at: "2026-04-12T08:00:00Z",
      from_focus: ["미용 시술"],
      to_focus: ["일반 진료 (아토피·여드름)"],
      reason: "feedback_accumulated",
      notes: "👎 피드백 18건 누적으로 재분류",
    },
    {
      changed_at: "2026-01-15T03:30:00Z",
      from_focus: [],
      to_focus: ["미용 시술"],
      reason: "scheduled_recrawl",
      notes: "초기 분류",
    },
  ],

  related_hospitals: [
    {
      hospital_id: "h_def456",
      name: "△△피부과",
      primary_focus: ["일반 진료 (아토피·여드름)"],
      similarity_score: 0.91,
      recommendation_type: "same_focus",
      distance_km: 0.8,
    },
    {
      hospital_id: "h_jkl012",
      name: "공덕가족피부과",
      primary_focus: ["일반 진료 (아토피)"],
      similarity_score: 0.84,
      recommendation_type: "same_focus",
      distance_km: 1.0,
    },
    {
      hospital_id: "h_ghi789",
      name: "□□피부과",
      primary_focus: ["사마귀·점 제거"],
      similarity_score: 0.42,
      recommendation_type: "fills_gap",
      distance_km: 1.2,
    },
    {
      hospital_id: "h_mno345",
      name: "마포모발의원",
      primary_focus: ["탈모 처방", "두피 관리"],
      similarity_score: 0.31,
      recommendation_type: "fills_gap",
      distance_km: 0.9,
    },
  ],

  metadata: {
    last_updated_at: "2026-04-12T08:00:00Z",
    data_sources: ["self_site", "public_registry", "user_reviews", "blog"],
    data_completeness: 0.82,
    warning: null,
  },
};
